pkg/webhook/resourceusage/handlers.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (r *ResourceUsageWatcher) onPodAdded(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	// A pod launched by the Spark operator will already be accounted for by the CRD informer callback
+	if !launchedBySparkOperator(pod.ObjectMeta) {
+		r.setResources("Pod", namespaceOrDefault(pod.ObjectMeta), pod.ObjectMeta.Name, podResourceUsage(pod), r.usageByNamespacePod)
+	}
+}
+
+func (r *ResourceUsageWatcher) onPodUpdated(oldObj, newObj interface{}) {
+	newPod := newObj.(*corev1.Pod)
+	if !launchedBySparkOperator(newPod.ObjectMeta) {
+		if newPod.Status.Phase == corev1.PodFailed || newPod.Status.Phase == corev1.PodSucceeded {
+			r.deleteResources("Pod", namespaceOrDefault(newPod.ObjectMeta), newPod.ObjectMeta.Name, r.usageByNamespacePod)
+		} else {
+			r.setResources("Pod", namespaceOrDefault(newPod.ObjectMeta), newPod.ObjectMeta.Name, podResourceUsage(newPod), r.usageByNamespacePod)
+		}
+	}
+}
+
+func (r *ResourceUsageWatcher) onPodDeleted(obj interface{}) {
+	var pod *corev1.Pod
+	switch o := obj.(type) {
+	case *corev1.Pod:
+		pod = o
+	case cache.DeletedFinalStateUnknown:
+		pod = o.Obj.(*corev1.Pod)
+	default:
+		return
+	}
+	if !launchedBySparkOperator(pod.ObjectMeta) {
+		r.deleteResources("Pod", namespaceOrDefault(pod.ObjectMeta), pod.ObjectMeta.Name, r.usageByNamespacePod)
+	}
+}
+
+func (r *ResourceUsageWatcher) onSparkApplicationAdded(obj interface{}) {
+	app := obj.(*so.SparkApplication)
+	resources, err := sparkApplicationResourceUsage(*app)
+	namespace := namespaceOrDefault(app.ObjectMeta)
+	if err != nil {
+		glog.Errorf("couldn't process SparkApplication %s/%s: %v", namespace, app.ObjectMeta.Name, err)
+	} else {
+		r.setResources("SparkApplication", namespace, app.ObjectMeta.Name, resources, r.usageByNamespaceApplication)
+	}
+}
+
+func (r *ResourceUsageWatcher) onSparkApplicationUpdated(oldObj, newObj interface{}) {
+	oldApp := oldObj.(*so.SparkApplication)
+	newApp := newObj.(*so.SparkApplication)
+	if oldApp.ResourceVersion == newApp.ResourceVersion {
+		return
+	}
+	namespace := namespaceOrDefault(newApp.ObjectMeta)
+	newResources, err := sparkApplicationResourceUsage(*newApp)
+	if err != nil {
+		glog.Errorf("couldn't process SparkApplication %s/%s: %v", namespace, newApp.ObjectMeta.Name, err)
+	} else {
+		r.setResources("SparkApplication", namespace, newApp.ObjectMeta.Name, newResources, r.usageByNamespaceApplication)
+	}
+}
+
+func (r *ResourceUsageWatcher) onSparkApplicationDeleted(obj interface{}) {
+	var app *so.SparkApplication
+	switch o := obj.(type) {
+	case *so.SparkApplication:
+		app = o
+	case cache.DeletedFinalStateUnknown:
+		app = o.Obj.(*so.SparkApplication)
+	default:
+		return
+	}
+	namespace := namespaceOrDefault(app.ObjectMeta)
+	r.deleteResources("SparkApplication", namespace, app.ObjectMeta.Name, r.usageByNamespaceApplication)
```

Comment:
Should use the constant `KindSparkApplication` instead of the string literal here. Ditto in other places.